### PR TITLE
feat: expressify `lower_bound` and `upper_bound` in `is_between`

### LINKED ISSUE #1659: [Bug]: is_between incorrect behaviour (TypeError: cannot create expression literal for value of type Expr)
### Describe the bug

When attempting to use nw.lit to create literal expressions for filtering dates in a Narwhal DataFrame, the following error occurs:

`TypeError: cannot create expression literal for value of type Expr.`

This happens specifically when using nw.lit in combination with `.filter()` and `.is_between()` for date-based comparisons. However, casting and aliasing the literals and then using them in `.to_native()` before filtering resolves the issue (but I can't do this since I want to have agnostic dataframe filtering.

### Steps or code to reproduce the bug

Setup:

```python
import polars as pl
import narwhal as nw

start_train_date = "2024-01-01"
end_train_date = "2024-03-05"

df_polars = pl.DataFrame({
    "application_started_at": [
        "2024-01-01", "2024-02-01", "2024-03-01", "2024-04-01", "2024-04-06",
    ]
})

df_native = nw.from_native(df_polars)
```

then:

```python
# This will raise the TypeError
df_native.select(
    nw.col("application_started_at").cast(nw.Date).dt.date(),
    nw.lit(start_train_date).cast(nw.Date).alias("a"),
    nw.lit(end_train_date).cast(nw.Date).alias("b")
).filter(
    nw.col("application_started_at").is_between(nw.col("a"), nw.col("b"), closed="both")
).to_native()
```

to get the error and:

```python
(
    df_native
    .select(
        nw.col(col_time).cast(nw.Date).dt.date(),
        nw.lit(start_train_date).cast(nw.Date).alias("a"),
        nw.lit(end_train_date).cast(nw.Date).alias("b")
    )
    .to_native()
    .filter(
        pl.col(col_time).is_between(pl.col("a"), pl.col("b"), closed="both")
    )
)
```

to get the expected results.

### Expected results

application_started_at | a | b
-- | -- | --
2024-01-01 | 2024-01-01 | 2024-03-05
2024-02-01 | 2024-01-01 | 2024-03-05
2024-03-01 | 2024-01-01 | 2024-03-05

### Actual results

`TypeError: cannot create expression literal for value of type Expr.`

### Please run narwhals.show_version() and enter the output below.

```shell
1.19.1
```

### Relevant log output

```shell
TypeError: cannot create expression literal for value of type Expr.

Hint: Pass `allow_object=True` to accept any value and create a literal of type Object.
File <command-332563390273393>, line 19
      4 df_polars = pl.DataFrame({
      5     "application_started_at": [
      6         "2024-01-01", "2024-02-01", "2024-03-01", "2024-04-01", "2024-04-06",
      7     ]
      8 })
     10 df_native = nw.from_native(df_polars)
     12 (
     13     df_native
     14     .select(
     15         nw.col(col_time).cast(nw.Date).dt.date(),
     16         nw.lit(start_train_date).cast(nw.Date).alias("a"),
     17         nw.lit(end_train_date).cast(nw.Date).alias("b")
     18     )
---> 19     .filter(
     20         nw.col(col_time).is_between(nw.col("a"), nw.col("b"), closed="both")
     21     )
     22     .to_native()
     23 )
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-ad765748-653a-475d-a18a-358020328a68/lib/python3.11/site-packages/polars/functions/lit.py:179, in lit(value, dtype, allow_object)
    176 except AttributeError:
    177     item = value
--> 179 return wrap_expr(plr.lit(item, allow_object, is_scalar=True))
```

### PR DESCRIPTION
**Pull request by MarcoGorelli**
## What type of PR is this? (check all applicable

closes #1659 

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues

- Related issue #\<issue number\>
- Closes #\<issue number\>

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added
- [ ] Documented the changes

### CODE REVIEW

File: narwhals/_arrow/namespace.py [hunk 1]
```diff
@@ -348,31 +348,6 @@ def concat(
...
-    def max(self: Self, *column_names: str) -> ArrowExpr:
-        return ArrowExpr.from_column_names(
-            *column_names, backend_version=self._backend_version, version=self._version
-        ).max()
-
-    def min(self: Self, *column_names: str) -> ArrowExpr:
-        return ArrowExpr.from_column_names(
-            *column_names, backend_version=self._backend_version, version=self._version
-        ).min()
-
```
Comment (MarcoGorelli) [inline narwhals/_arrow/namespace.py:375]: drive-by - it's redundat to define all of these in the `CompliantExpr`s

Comment (FBruzzesi) [reply narwhals/_arrow/namespace.py:375]: Nice one! I love to see net negative in PRs 👌

File: tests/expr_and_series/is_between_test.py [hunk 2]
```diff
@@ -22,12 +18,21 @@
...
     df = nw.from_native(constructor(data))
     result = df.select(nw.col("a").is_between(1, 5, closed=closed))
     expected_dict = {"a": expected}
     assert_equal_data(result, expected_dict)
 
 
+def test_is_between_expressified(constructor: Constructor) -> None:
+    data = {"a": [1, 4, 2, 5], "b": [0, 5, 2, 4], "c": [9, 9, 9, 9]}
+    df = nw.from_native(constructor(data))
+    result = df.select(nw.col("a").is_between(nw.col("b"), nw.col("c")))
```
Comment (FBruzzesi) [inline tests/expr_and_series/is_between_test.py:31]: Should we test with a generic expression instead of _just_ `col`?
Random proposal that would not change `expected_dict`:

```suggestion
    result = df.select(nw.col("a").is_between(nw.col("b") * 0.9, nw.col("c") - 1))
```

**Review by FBruzzesi — COMMENTED** [review]
Left a minor comment on the unit test.

The approach used here to expressify `is_between` seems generic enough to be used elsewhere, is that right?! I somehow thought it was going to be fairly more complex, yet `extract_compliant` together with `<>_and_extract_native` seem all we need 🙌🏼🚀

**Comment by FBruzzesi** [comment]
On a second thought, we are creating some asymmetry between what we can pass to `Series.is_between` vs `Expr.is_between`, aren't we? I would expect to be able to pass series' to  `Series.is_between`, e.g.

```py
some_series.is_between(lower_bound_series, upper_bound_series)
``` 

Apologies if this is already possible, if so we may want to add a test case for it

**Comment by MarcoGorelli** [comment]
thanks for your review! yup, this is what we do in `__eq__` to accept either scalars or expressions / series, we can almost certainly do it in more places (and probably reuse more code)
